In [9]:
from datetime import date
import pandas as pd

pd.set_option('display.max_columns', 350)

In [10]:
import os
def get_project_root_path():
    project_directory = "fifa22-players-analysis"
    separated_path = os.getcwd().split(os.sep)
    project_directory_index = separated_path.index(project_directory) + 1
    return os.sep.join(separated_path[0:project_directory_index])


import joblib
def save_model(model, model_name):
    filename = f"model_name_{model_name}_{str(date.today())}.pkl"
    print(filename)
    model_path = os.sep.join([get_project_root_path(), "saved_models", filename])
    joblib.dump(model, model_path)

In [11]:
data = pd.read_csv(f"data/processed_data_{date.today()}.tsv", sep="\t")
data.head()

<ipython-input-11-0e18a27736b9>:1: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv(f"data/processed_data_2022-06-25.tsv", sep="\t")


sofifa_id                                         player_url  short_name  \
0     167575  https://sofifa.com/player/167575/anton-shunin/...   A. Shunin   
1     205211  https://sofifa.com/player/205211/milan-borjan/...   M. Borjan   
2     244700  https://sofifa.com/player/244700/yuriy-dyupin/...   Y. Dyupin   
3     199904  https://sofifa.com/player/199904/jesse-joronen...  J. Joronen   
4     240225  https://sofifa.com/player/240225/matvey-safono...  M. Safonov   

             long_name player_positions  overall  potential  value_eur  \
0         Anton Shunin               GK       76         76        NaN   
1         Milan Borjan               GK       76         76        NaN   
2         Yuriy Dyupin               GK       74         74        NaN   
3  Jesse Pekka Joronen               GK       72         72        NaN   
4       Matvey Safonov               GK       72         82        NaN   

   wage_eur         dob  weight_kg  height_cm  club_team_id club_position  \
0       NaN  1987-01-27         88        191           NaN           NaN   
1       NaN  1987-10-23         84        196           NaN           NaN   
2       NaN  1988-03-17         84        187           NaN           NaN   
3       NaN  1993-03-21         89        197           NaN           NaN   
4       NaN  1999-02-25         82        192           NaN           NaN   

   club_jersey_number preferred_foot  weak_foot  skill_moves  \
0                 NaN           Left          3            1   
1                 NaN          Right          2            1   
2                 NaN          Right          3            1   
3                 NaN          Right          3            1   
4                 NaN          Right          2            1   

   international_reputation      work_rate      body_type  release_clause_eur  \
0                         1  Medium/Medium    Lean (185+)                 NaN   
1                         1  Medium/Medium  Normal (185+)                 NaN   
2                         1  Medium/Medium  Normal (185+)                 NaN   
3                         1  Medium/Medium  Normal (185+)                 NaN   
4                         1  Medium/Medium  Normal (185+)                 NaN   

  player_tags                         player_traits  pace  shooting  passing  \
0         NaN                           Team Player   NaN       NaN      NaN   
1         NaN         Leadership, Comes For Crosses   NaN       NaN      NaN   
2         NaN                                   NaN   NaN       NaN      NaN   
3         NaN  GK Long Throw, Cautious With Crosses   NaN       NaN      NaN   
4         NaN                                   NaN   NaN       NaN      NaN   

   dribbling  defending  physic  attacking_crossing  attacking_finishing  \
0        NaN        NaN     NaN                  18                   11   
1        NaN        NaN     NaN                  15                   12   
2        NaN        NaN     NaN                  12                    6   
3        NaN        NaN     NaN                  15                   13   
4        NaN        NaN     NaN                  13                    7   

   attacking_heading_accuracy  skill_dribbling  skill_curve  \
0                          12               14           15   
1                          24               28           16   
2                          10                7           14   
3                          18               16           14   
4                          14               10           11   

   skill_fk_accuracy  skill_long_passing  skill_ball_control  \
0                 13                  30                  20   
1                 16                  48                  23   
2                 11                  22                  12   
3                 12                  25                  16   
4                 12                  26                  18   

   movement_acceleration  movement_sprint_spee

In [12]:
def count_na(df: pd.DataFrame) -> None:
    nan_sum = df.isna().sum()
    display(nan_sum[nan_sum > 0])

count_na(data)

value_eur                74
wage_eur                 61
club_team_id             61
club_position            61
club_jersey_number       61
release_clause_eur     1176
player_tags           17798
player_traits          9841
pace                   2132
shooting               2132
passing                2132
dribbling              2132
defending              2132
physic                 2132
goalkeeping_speed     17107
BMI                   19234
obese                 19234
lean                  19234
normal                19234
dtype: int64

Analisando essa tabela com os valores nulos podemos perceber que alguns jogadores do dataset não possuem um time atual, isso explica
`wage_eur`, `club_team_id`, `club_position`, `club_jersey_number` terem a mesma quantidade de valores nulos (61),
porém `value_eur` deveria seguir esse mesmo padrão, assim podemos dizer que 13 (74 - 61) jogadores que estão ativos não possuem um valor,
portanto precisamos prever esses dados.

Também podemos perceber que temos 2132 jogadores não possuem `pace`, `shooting`, `passing`, `dribbling`, `defending`, `physic`. Vamos analisar esses jogadores

In [13]:
def percent_of_gk(df: pd.DataFrame) -> None:
    gk = df[df['GK'] == 1]
    display(gk.shape[0] / df.shape[0])


missing_main_attr_data = data[data.pace.isna()]

percent_of_gk(missing_main_attr_data)
missing_main_attr_data

1.0

sofifa_id                                         player_url  \
0        167575  https://sofifa.com/player/167575/anton-shunin/...   
1        205211  https://sofifa.com/player/205211/milan-borjan/...   
2        244700  https://sofifa.com/player/244700/yuriy-dyupin/...   
3        199904  https://sofifa.com/player/199904/jesse-joronen...   
4        240225  https://sofifa.com/player/240225/matvey-safono...   
...         ...                                                ...   
2127     257618  https://sofifa.com/player/257618/armin-gremsl/...   
2128     262424  https://sofifa.com/player/262424/sorin-mogosan...   
2129     262420  https://sofifa.com/player/262420/robert-popa/2...   
2130     210022  https://sofifa.com/player/210022/philip-sprint...   
2131     244417  https://sofifa.com/player/244417/julian-krahl/...   

       short_name            long_name player_positions  overall  potential  \
0       A. Shunin         Anton Shunin               GK       76         76   
1       M. Borjan         Milan Borjan               GK       76         76   
2       Y. Dyupin         Yuriy Dyupin               GK       74         74   
3      J. Joronen  Jesse Pekka Joronen               GK       72         72   
4      M. Safonov       Matvey Safonov               GK       72         82   
...           ...                  ...              ...      ...        ...   
2127    A. Gremsl         Armin Gremsl               GK       63         66   
2128  S. Mogoșanu       Sorin Mogoșanu               GK       58         60   
2129      R. Popa          Robert Popa               GK       56         75   
2130    P. Sprint        Philip Sprint               GK       62         64   
2131     J. Krahl         Julian Krahl               GK       62         76   

      value_eur  wage_eur         dob  weight_kg  height_cm  club_team_id  \
0           NaN       NaN  1987-01-27         88        191           NaN   
1           NaN       NaN  1987-10-23         84        196           NaN   
2           NaN       NaN  1988-03-17         84        187           NaN   
3           NaN       NaN  1993-03-21         89        197           NaN   
4           NaN       NaN  1999-02-25         82        192           NaN   
...         ...       ...         ...        ...        ...           ...   
2127   475000.0    2000.0  1994-08-13         85        190      115716.0   
2128   170000.0    1000.0  1992-11-22         77        180      115716.0   
2129   325000.0     500.0  2003-03-05         83        190      115716.0   
2130   350000.0     700.0  1993-06-27         98        196      115820.0   
2131   850000.0     500.0  2000-01-22         90        194      115820.0   

     club_position  club_jersey_number preferred_foot  weak_foot  skill_moves  \
0              NaN                 NaN           Left          3            1   
1              NaN                 NaN          Right          2            1   
2              NaN                 NaN          Right          3            1   
3              NaN                 NaN          Right          3            1   
4              NaN                 NaN          Right          2            1   
...            ...                 ...            ...        ...          ...   
2127           RES                94.0           Left          2            1   
2128            GK                 1.0          Right          1            1   
2129           SUB                31.0          Right          2            1   
2130            GK                 1.0          Right          2            1   
2131           SUB                13.0          Right          3            1   

      international_reputation      work_rate         body_type  \
0                            1  Medium/Medium       Lean (185+)   
1                            1  Medium/Medium     Normal (185+)   
2                            1  Medium/Medium     Normal (185+)   
3                            1  Medium/Medium     

Como vimos que os goleiros não tem os atributos listados acima, mas possuem um overall em cada posição, seria interessante prever esses atributos.

In [14]:
missing_gk_speed_data = data[data.goalkeeping_speed.isna()]
sum_of_missing_data = missing_gk_speed_data.shape[0] + missing_main_attr_data.shape[0]
print(f"Jogadores sem gk speed + jogadores sem atributos primários: {sum_of_missing_data}")
print(f"Numero de jogadores: {data.shape[0]}")

Jogadores sem gk speed + jogadores sem atributos primários: 19239
Numero de jogadores: 19239


Assim podemos ver que todos os jogadores que não são goleiros não possuem `goalkeeping_speed`

## Prevendo valor dos jogadores
Nessa etapa vamos criar um modelo para prever o valor dos 13 jogadores que não possuem valor

In [27]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.feature_selection import RFECV
from sklearn.impute import SimpleImputer
import numpy as np

def predict_and_populate(df_to_populate, df_to_train, model, column_to_predict, round_=None, save=False) -> pd.DataFrame:
    # Separating the target variable
    target = df_to_train[column_to_predict]
    features = df_to_train.drop(column_to_predict, axis=1)

    # Poluting the missing features
    mean_imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
    mean_imputer.fit(features)
    imputed_features = mean_imputer.transform(features)

    # Train test split
    features_train, features_test, target_train, target_test = train_test_split(imputed_features, target, test_size=0.2)

    # Training the model
    trained_model = model.fit(features_train, target_train)

    # Saving the model
    if save:
        model_name = str(model).split("(")[0]
        filename = f"{model_name}_{column_to_predict}"
        save_model(trained_model, filename)

    # Calculating model score
    model_score = trained_model.score(features_test, target_test)
    print(f"Score do R² modelo: {model_score:.4f}")


    if model == LinearRegression or DummyRegressor or RandomForestRegressor:
        mse = mean_squared_error(target_test, trained_model.predict(features_test))
        mae = mean_absolute_error(target_test, trained_model.predict(features_test))
        print(f"MSE: {mse:.4f}")
        print(f"MAE: {mae:.4f}")

    # Populating the missing values
    feature_to_populate = mean_imputer.transform(df_to_populate.drop(column_to_predict, axis=1))
    predicted_values = trained_model.predict(feature_to_populate)


    if round_ is not None:
        predicted_values = predicted_values.round(round_)

    df_to_populate[column_to_predict] = predicted_values

    return df_to_populate




In [28]:
players_with_club_and_no_value = data.query("not club_team_id.isna() and value_eur.isna()").copy()
data_without_players_with_club_and_no_value = data.drop(players_with_club_and_no_value.index)

data_to_train = data_without_players_with_club_and_no_value.query("not value_eur.isna()").copy()
columns_to_drop = [
    "sofifa_id", "player_url", "short_name", "long_name",
    "dob", "club_position","player_traits", "player_tags",
    "player_positions", "work_rate", "body_type",
    "preferred_foot", "nationality_name"
]

data_to_train = data_to_train.drop(columns=columns_to_drop)
data_to_fill = players_with_club_and_no_value.drop(columns=columns_to_drop)

In [29]:
# Normalizando os dados
# from sklearn.preprocessing import StandardScaler
#
# standerd_scaler =  StandardScaler()
# standardized_features = standerd_scaler.fit_transform(imputed_features)
# standardized_features

## Comparando modelos diferentes

In [30]:
# TODO ver tanto o MSE quanto o MAE
# TODO aplicar RFECV em cada modelo

from sklearn.dummy import DummyRegressor
dummy = DummyRegressor()
predict_and_populate(data_to_fill, data_to_train, dummy, "value_eur")

Score do R² modelo: -0.0005
MSE: 67637941553168.6250
MAE: 3290389.7069


overall  potential     value_eur  wage_eur  weight_kg  height_cm  \
11929       69         69  2.814726e+06    3000.0         68        169   
11987       67         67  2.814726e+06    3000.0         80        186   
12264       69         69  2.814726e+06    5000.0         72        174   
13149       75         75  2.814726e+06   13000.0         79        182   
13177       73         73  2.814726e+06   10000.0         74        173   
14103       56         56  2.814726e+06    1000.0         89        188   
14611       67         67  2.814726e+06     500.0         76        180   
15125       75         75  2.814726e+06     900.0         69        174   
16065       67         67  2.814726e+06    5000.0         75        180   
16498       58         58  2.814726e+06    2000.0         76        185   
16814       58         58  2.814726e+06     800.0         76        175   
18159       63         63  2.814726e+06     800.0         71        178   
18166       59         59  2.814726e+06     700.0         72        177   

       club_team_id  club_jersey_number  weak_foot  skill_moves  \
11929      100851.0                 7.0          3            3   
11987      101006.0                38.0          2            2   
12264      101084.0                25.0          4            3   
13149      110395.0                 9.0          4            3   
13177      110396.0                11.0          4            4   
14103      110799.0                14.0          3            2   
14611      111007.0                 3.0          3            2   
15125      111325.0                10.0          3            4   
16065      111839.0                10.0          4            3   
16498      112254.0                22.0          3            2   
16814      112444.0                44.0          4            3   
18159      113197.0                10.0          3            3   
18166      113197.0                11.0          4            3   

       international_reputation  release_clause_eur  pace  shooting  passing  \
11929                         2                 NaN  70.0      70.0     61.0   
11987                         1                 NaN  32.0      43.0     48.0   
12264                         1                 NaN  64.0      61.0     65.0   
13149                         1                 NaN  51.0      79.0     62.0   
13177                         3                 NaN  64.0      74.0     76.0   
14103                         1                 NaN  50.0      56.0     39.0   
14611                         1                 NaN  30.0      35.0     46.0   
15125                         1                 NaN  42.0      71.0     80.0   
16065                         1                 NaN  35.0      64.0     67.0   
16498                         1                 NaN  42.0      63.0     59.0   
16814                         2                 NaN  30.0      55.0     62.0   
18159                         3                 NaN  31.0      67.0     74.0   
18166                         1                 NaN  30.0      63.0     55.0   

       dribbling  defending  physic  attacking_crossing  attacking_finishing  \
11929       72.0       29.0    61.0                  60                   71   
11987       48.0       68.0    74.0                  44                   30   
12264       72.0       68.0    66.0                  80                   52   
13149       68.0       37.0    70.0                  53                   81   
13177       75.0       44.0    64.0                  72                   73   
14103       53.0       23.0    63.0                  34                   58   
14611       50.0       67.0    71.0                  35                   34   
15125       75.0       38.0    50.0                  79                   69   
16065       66.0       67.0    59.0                  61                   60   
16498       62.0       43.0    68.0                  61                   60   
16814       59.0

In [31]:
from sklearn.linear_model import LinearRegression

linear_regression = LinearRegression()
predict_and_populate(data_to_fill, data_to_train, linear_regression, "value_eur")

Score do R² modelo: 0.9734
MSE: 1456919770884.8223
MAE: 499518.8803


overall  potential     value_eur  wage_eur  weight_kg  height_cm  \
11929       69         69  2.645781e+06    3000.0         68        169   
11987       67         67  1.533158e+06    3000.0         80        186   
12264       69         69  2.083065e+06    5000.0         72        174   
13149       75         75  2.724606e+06   13000.0         79        182   
13177       73         73  3.749650e+06   10000.0         74        173   
14103       56         56  1.053921e+06    1000.0         89        188   
14611       67         67  1.718336e+06     500.0         76        180   
15125       75         75  1.966312e+06     900.0         69        174   
16065       67         67  1.900290e+06    5000.0         75        180   
16498       58         58  1.181943e+06    2000.0         76        185   
16814       58         58  1.971310e+06     800.0         76        175   
18159       63         63  4.575668e+06     800.0         71        178   
18166       59         59  6.617684e+05     700.0         72        177   

       club_team_id  club_jersey_number  weak_foot  skill_moves  \
11929      100851.0                 7.0          3            3   
11987      101006.0                38.0          2            2   
12264      101084.0                25.0          4            3   
13149      110395.0                 9.0          4            3   
13177      110396.0                11.0          4            4   
14103      110799.0                14.0          3            2   
14611      111007.0                 3.0          3            2   
15125      111325.0                10.0          3            4   
16065      111839.0                10.0          4            3   
16498      112254.0                22.0          3            2   
16814      112444.0                44.0          4            3   
18159      113197.0                10.0          3            3   
18166      113197.0                11.0          4            3   

       international_reputation  release_clause_eur  pace  shooting  passing  \
11929                         2                 NaN  70.0      70.0     61.0   
11987                         1                 NaN  32.0      43.0     48.0   
12264                         1                 NaN  64.0      61.0     65.0   
13149                         1                 NaN  51.0      79.0     62.0   
13177                         3                 NaN  64.0      74.0     76.0   
14103                         1                 NaN  50.0      56.0     39.0   
14611                         1                 NaN  30.0      35.0     46.0   
15125                         1                 NaN  42.0      71.0     80.0   
16065                         1                 NaN  35.0      64.0     67.0   
16498                         1                 NaN  42.0      63.0     59.0   
16814                         2                 NaN  30.0      55.0     62.0   
18159                         3                 NaN  31.0      67.0     74.0   
18166                         1                 NaN  30.0      63.0     55.0   

       dribbling  defending  physic  attacking_crossing  attacking_finishing  \
11929       72.0       29.0    61.0                  60                   71   
11987       48.0       68.0    74.0                  44                   30   
12264       72.0       68.0    66.0                  80                   52   
13149       68.0       37.0    70.0                  53                   81   
13177       75.0       44.0    64.0                  72                   73   
14103       53.0       23.0    63.0                  34                   58   
14611       50.0       67.0    71.0                  35                   34   
15125       75.0       38.0    50.0                  79                   69   
16065       66.0       67.0    59.0                  61                   60   
16498       62.0       43.0    68.0                  61                   60   
16814       59.0

In [32]:
from sklearn.ensemble import RandomForestRegressor

random_forest = RandomForestRegressor(random_state=0, n_jobs=-1, oob_score=True)
players_with_club_and_no_value = predict_and_populate(
    data_to_fill, data_to_train, random_forest, "value_eur", save=True
);

model_name_RandomForestRegressor_value_eur_2022-06-27.pkl
Score do R² modelo: 0.9917
MSE: 495300835067.9363
MAE: 169891.9123


### Preenchendo os dados a partir do modelo criado

In [ ]:
data.loc[players_with_club_and_no_value.index, "value_eur"] = players_with_club_and_no_value.value_eur
data.loc[players_with_club_and_no_value.index]

In [ ]:
count_na(data)

In [ ]:
# Removendo jogadores não ativos
data = data.query("not club_team_id.isna()").copy()

count_na(data)

# Prevendo pace, shooting, passing, dribbling, defending, physic dos goleiros

In [ ]:
gk_data = data.query("GK == 1").copy()
data_without_gk = data.query("GK == 0").copy()

columns_to_predict = ["pace", "shooting", "passing", "dribbling", "defending", "physic"]

for column_to_predict in columns_to_predict:

    columns_to_drop_with_other_atributes = [col for col in columns_to_predict if col != column_to_predict]
    columns_to_drop_with_other_atributes.extend(columns_to_drop)

    data_to_train = data_without_gk.drop(columns=columns_to_drop_with_other_atributes)
    data_to_predict = gk_data.drop(columns=columns_to_drop_with_other_atributes)

    linear_regression = LinearRegression()
    print(f"Prevendo {column_to_predict}")
    filled_data = predict_and_populate(
        data_to_predict, data_to_train, linear_regression,
        column_to_predict, 0, save=True
    )
    print()

    data.loc[filled_data.index, column_to_predict] = filled_data[column_to_predict]

# data.loc[gk_data.index]

In [ ]:
count_na(data)

## Prevendo goalkeeping_speed dos jogadores

In [ ]:
players_data = data.query("GK == 0").copy()
players_data

In [ ]:
data_to_predict = players_data.drop(columns=columns_to_drop)
data_to_train = gk_data.drop(columns=columns_to_drop)

column_to_predict = "goalkeeping_speed"

linear_regression = LinearRegression()
filled_data = predict_and_populate(
    data_to_predict, data_to_train, linear_regression, column_to_predict, 0, save=True
)

data.loc[filled_data.index, column_to_predict] = filled_data[column_to_predict]

In [ ]:
count_na(data)

In [ ]:
from datetime import date
data.drop(columns=["release_clause_eur"], inplace=True)

filled_data_save_path = os.sep.join([get_project_root_path(), "preprocessing", "data", f"filled_data_{date.today()}.tsv"])
data.to_csv(filled_data_save_path, index=False, sep="\t")
filled_data_save_path

In [ ]:
count_na(data)

In [ ]:
data.describe().T.round()